### ISSUE TITLE
Issue using ttnn.typecast to convert shape

### ISSUE THREAD

**Issue by cmaryanTT** [tenstorrent/tt-metal#35487]
From: https://tenstorrent.atlassian.net/servicedesk/customer/portal/1/CUST-603

**Software Stack**
TT-Metallium

**Hardware Platform**
6U Galaxy - WH

**Description**
(Copied from slack channel)
We were working on Galaxy machine we have, and met below issue.
Could you check and figure out the cause of this issue, and related solution?

**Issue**
when trying to use ttnn.typecast to convert shape (32,256,4096) bfloat16 tensor of prefill progress
to bfloat8_b, we encountered below issue (continue to following thread) :

```
==================================================
Testing Device 5
==================================================
2025-12-22 21:10:48.367 | info     |          Fabric | TopologyMapper mapping start (mesh=0): n_log=32, n_phys=32, log_deg_hist={2:4, 3:16, 4:12}, phys_deg_hist={4:32} (topology_mapper.cpp:574)
2025-12-22 21:10:48.642 | warning  |           Metal | Opening subset of mmio devices slows down UMD read/write to remote chips. If opening more devices, consider using CreateDevices API. (device_pool.cpp:303)
2025-12-22 21:10:48.643 | info     |           Metal | Profiler started on device 5 (device_pool.cpp:203)
Max difference: 2.0
Number of different elements: 40
Different values: tensor([2.0000, 2.0000, 0.0156, 0.0156, 0.0156, 0.0156, 2.0000, 0.0156, 0.0156,
        0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 2.0000, 0.0156,
        0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 0.0156, 2.0000,
        0.0156, 0.0156, 0.0156, 2.0000, 0.0156, 0.0156, 2.0000, 0.0156, 0.0156,
        0.0156, 0.0156, 0.0156, 2.0000], dtype=torch.bfloat16)
Positions of differences: (tensor([ 1,  5,  6,  6,  6,  6,  6,  6,  6,  9,  9,  9,  9,  9,  9,  9,  9,  9,
        17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 20, 20, 20, 20, 20, 26, 28, 28,
        28, 28, 28, 28]), tensor([160, 218,  67,  67,  67,  67,  67,  67,  67, 241, 241, 241, 241, 241,
        241, 241, 241, 241, 198, 198, 198, 198, 198, 198, 198, 198, 198, 198,
         30,  30,  30,  30,  30,  34,   2,   2,   2,   2,   2,   2]), tensor([2862, 1322,  288,  292,  293,  294,  295,  299,  301, 1332, 1333, 1334,
        1335, 1336, 1337, 1338, 1340, 1341, 4018, 4019, 4021, 4022, 4023, 4025,
        4026, 4027, 4028, 4031,  544,  545,  546,  548,  553, 3371, 1056, 1060,
        1061, 1062, 1065, 1066]))
:x: Device 5 FAILED!
```

Error reproduction Script:

```
# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC.
# SPDX-License-Identifier: Apache-2.0
"""
Test for bfloat16 -> bfloat8_b typecast consistency across all devices.
This test checks if the same typecast operation produces identical results
when run twice on the same input tensor, for each device individually.
"""
import torch
import ttnn
def as_torch(tensor):
    """Convert ttnn tensor to torch tensor."""
    return ttnn.to_torch(tensor, dtype=torch.bfloat16)
def test_typecast_bfloat16_to_bfloat8_all_devices():
    """
    Test that bfloat16 -> bfloat8_b typecast produces consistent results
    on all available devices.
    Iterates through all devices and performs the typecast operation twice
    on the same input, checking if results are identical.
    """
    torch.manual_seed(42)
    shape = (32, 256, 4096)
    # Create random input tensor in bfloat16
    input_data = torch.randn(shape, dtype=torch.bfloat16)
    # Get all available device IDs
    num_devices = ttnn.GetNumAvailableDevices()
    print(f"\nNumber of available devices: {num_devices}")
    failed_devices = []
    for device_id in range(num_devices):
        print(f"\n{'='*50}")
        print(f"Testing Device {device_id}")
        print(f"{'='*50}")
        # Open device
        device = ttnn.open_device(device_id=device_id)
        try:
            # Convert to ttnn tensor on this device
            input_tensor = ttnn.from_torch(
                input_data,
                dtype=ttnn.bfloat16,
                layout=ttnn.TILE_LAYOUT,
                device=device,
                memory_config=ttnn.DRAM_MEMORY_CONFIG,
            )
            # Perform typecast twice
            output_1 = ttnn.typecast(input_tensor, ttnn.bfloat4_b)
            output_2 = ttnn.typecast(input_tensor, ttnn.bfloat4_b)
            # Convert back to torch for comparison
            result_1 = as_torch(output_1)
            result_2 = as_torch(output_2)
            # Calculate difference
            diff = torch.abs(result_1 - result_2)
            max_diff = diff.max().item()
            num_diffs = (diff > 0).sum().item()
            print(f"Max difference: {max_diff}")
            print(f"Number of different elements: {num_diffs}")
            if num_diffs > 0:
                print(f"Different values: {diff[diff > 0]}")
                print(f"Positions of differences: {torch.where(diff > 0)}")
                failed_devices.append(device_id)
                print(f"❌ Device {device_id} FAILED!")
            else:
                print(f"✓ Device {device_id} PASSED!")
        finally:
            # Always close device
            ttnn.close_device(device)
    # Summary
    print(f"\n{'='*50}")
    print("SUMMARY")
    print(f"{'='*50}")
    print(f"Total devices tested: {num_devices}")
    print(f"Passed: {num_devices - len(failed_devices)}")
    print(f"Failed: {len(failed_devices)}")
    if failed_devices:
        print(f"Failed device IDs: {failed_devices}")
    assert len(failed_devices) == 0, f"Typecast produced different results on devices: {failed_devices}"
if __name__ == "__main__":
    test_typecast_bfloat16_to_bfloat8_all_devices()
```

Only one Galaxy machine of 4 Galaxy had this issue. We did this error reproduction just after power cycle, but had same result.

**Comment by zzigler-tt** [comment]
@ihnpark551 for visibility.

**Comment by mateusznowakTT** [comment]
Trying to reproduce, on 1 WH in a loop so far no luck, Galaxy is WIP.

Things to check once reproduced:
- Whether TT_METAL_SLOW_DISPATCH_MODE=true still reproduce the issue
- The same with explicit synchronization after CPU to device to CPU data transfers and operation
- If dropping bf16 conversion on CPU (ttnn.Tensor.to_torch returns fp32 tensor by default for bfp8 and bfp4, which is then converted to bf16 on CPU) changes the observed behavior
- Whether being more explicit in setting copy_tile and packer configuration in typecast kernel improves the situation

**Comment by ihnpark551** [comment]
Added this ticket to Moreh board, as P1 priority.

**Comment by mateusznowakTT** [comment]
Another to check once reproduced:
- If setting dst_full_sync_en = True changes observed behavior

**Comment by ihnpark551** [comment]
Hi @mateusznowakTT , thanks for your comment.
Because this error always occurs same device (device 5), therefore we need to check whether it is H/W issue or not.
We would firstly checking it, and will get back to you once the cause of the issue specified on H/W side.

**Comment by mateusznowakTT** [comment]
Hi @ihnpark551, yes, this seems like H/W issue, but I'm also trying to increase confidence there is no gap in SW.
So far, I hadn't been able to reproduce it on my side on Galaxy - I'm looking forward to hear once you have any findings.

**Comment by ihnpark551** [comment]
Hi @mateusznowakTT, I fully acknowledged the purpose of this ticket.
Once we check firstly on H/W side, and whether the issue has been solved or not, I would get back to here for validating on S/W side. Thanks for your support!

**Comment by zzigler-tt** [comment]
Removing SW P0 label as this is currently being triaged as a HW issue. 
If this is turns into a SW item we will re-evaluate priority.
